plugin/storage/badger/spanstore/reader.go

```diff
@@ -612,3 +624,57 @@ func scanRangeFunction(it *badger.Iterator, indexEndValue []byte) bool {
 	}
 	return false
 }
+
+func (r *TraceReader) loadServices() {
+	r.store.View(func(txn *badger.Txn) error {
+		opts := badger.DefaultIteratorOptions
+		it := txn.NewIterator(opts)
+		defer it.Close()
+
+		serviceKey := []byte{serviceNameIndexKey}
+
+		// Seek all the services first
+		for it.Seek(serviceKey); it.ValidForPrefix(serviceKey); it.Next() {
+			timestampStartIndex := len(it.Item().Key()) - (sizeOfTraceID + 8) // 8 = sizeof(uint64)
+			serviceName := string(it.Item().Key()[len(serviceKey):timestampStartIndex])
+			keyTTL := it.Item().ExpiresAt()
+			if v, found := r.cache.services[serviceName]; found {
+				if v > keyTTL {
+					continue
+				}
+			}
+			r.cache.services[serviceName] = keyTTL
+		}
+		return nil
+	})
+}
+
+func (r *TraceReader) loadOperations(service string) {
+	r.store.View(func(txn *badger.Txn) error {
+		opts := badger.DefaultIteratorOptions
+		it := txn.NewIterator(opts)
+		defer it.Close()
+
+		serviceKey := make([]byte, len(service)+1)
+		serviceKey[0] = operationNameIndexKey
+		copy(serviceKey[1:], service)
+
+		// Seek all the services first
+		for it.Seek(serviceKey); it.ValidForPrefix(serviceKey); it.Next() {
+			timestampStartIndex := len(it.Item().Key()) - (sizeOfTraceID + 8) // 8 = sizeof(uint64)
+			operationName := string(it.Item().Key()[len(serviceKey):timestampStartIndex])
+			keyTTL := it.Item().ExpiresAt()
+			if _, found := r.cache.operations[service]; !found {
+				r.cache.operations[service] = make(map[string]uint64)
+			}
+
+			if v, found := r.cache.operations[service][operationName]; found {
+				if v > keyTTL {
+					continue
+				}
+			}
+			r.cache.operations[service][operationName] = keyTTL
```

Comment:
you now went the other way - reader knows too much about the cache. There should be a clear API boundary between reader and cache, e.g. all the reader should be doing here is something like `cache.add(service, operation)`. Just because the cache is in the same package doesn't mean the reader should be accessing its private fields like that.